examples/experimental/dagster-airlift/dagster_airlift/core/dag_defs.py

```diff
@@ -151,9 +154,37 @@ def resolve_defs(
             )
         )
 
-    return Definitions.merge(
-        *list(
-            dag_def.to_defs() if isinstance(dag_def, DagDefs) else dag_def for dag_def in dag_defs
-        ),
-        *transformed_defs,
+    return coerce_specs_in_defs(
+        Definitions.merge(
+            *list(
+                dag_def.to_defs() if isinstance(dag_def, DagDefs) else dag_def
+                for dag_def in dag_defs
+            ),
+            *transformed_defs,
+        )
+    )
+
+
+def coerce_specs_in_defs(defs: Definitions) -> Definitions:
+    new_assets = []
+    for asset in defs.assets or []:
+        if isinstance(asset, AssetsDefinition):
+            new_assets.append(asset)
+        elif isinstance(asset, SourceAsset):
+            check.failed("SourceAsset not supported in this context")
+        elif isinstance(asset, CacheableAssetsDefinition):
+            check.failed("CacheableAssetsDefinition not supported in this context")
+        elif isinstance(asset, AssetSpec):
+            new_assets.append(external_asset_from_spec(asset))
+        else:
+            check.failed(f"Unsupported asset type {type(asset)}")
+    return Definitions(
+        assets=new_assets,
+        resources=defs.resources,
+        sensors=defs.sensors,
+        schedules=defs.schedules,
+        jobs=defs.jobs,
+        loggers=defs.loggers,
+        executor=defs.executor,
+        asset_checks=defs.asset_checks,
```

Comment:
so at this point I am confused as to what is going on here.  

This might be a problem with the core `Definitions` abstraction to be honest. It is serving two purposes right now. One is for a vessel for passed in things, which are unchanged. The other is the completely canonicalized representation. Different sets of properties are used in different contexts, which is very confusing.